File: Algo JS/circularlinkedlist.js
/***********************************************************************
 * LinkedList class - Circular linkedlist, adds item to the head of list
************************************************************************/
const Node = require('./node');

// colors for console error printout
const BACKGROUND_COLOR =    '\x1b[41m';            // red
const FOREGROUND_COLOR =    '\x1b[37m';            // white
const RESET =               '\x1b[0m';             // reset color

module.exports = class LinkedList {
    constructor()
    {
        this._head = new Node();
        
        // circular in nature
        this._head.previous = this._head;
        this._head.next = this._head;
        
        this._length = 0;
    }
    
    get length()
    {
        return this._length;
    }
    
    isEmpty()
    {
        return (this.length === 0);
    }
    
    has(value)
    {
        let iter = this._head.next;
        
        while (iter !== this._head)
        {
            if (iter.data === value)
            {
                return true;
            }
            
            iter = iter.next;
        }
        
        return false;
    }
    
    clear()
    {
        let iter = this._head.next;
        
        // iterate through the list
        while (iter !== this._head)
        {
            let node = iter;
            
            iter = iter.next;
            node = null;
            this._length -= 1;
        }
        
        // point to oneself
        this._head.previous = this._head;
        this._head.next = this._head;
    }
    
    _find(value)
    {
        let iter = this._head.next;
        
        while (iter !== this._head)
        {
            // item found
            if (iter.data === value)
            {
                return iter;
            }
            
            iter = iter.next;
        }
        
        // item not found
        return null;
    }
    
    get(index)
    {
        let iter = this._head.next;
        let i = 0;
        
        while (iter !== this._head)
        {
            if (i === index)
            {
                return iter.data;
            }
            
            i += 1;
            iter = iter.next;
        }
        
        return null;
    }
    
    set(index, value)
    {
        let iter = this._head.next;
        let i = 0;
        
        while (iter !== this._head)
        {
            if (i === index)
            {
                iter.data = value;
                return true;
            }
            
            i += 1;
            iter = iter.next;
        }
        
        return false;
    }
    
    insert(index, value)
    {
        // bound check
        if ((index < 0) || (index >= this._length)) 
        {
            console.log(BACKGROUND_COLOR, FOREGROUND_COLOR, "Index out of bounds!!!", RESET);
            
            return false;
        }
        
        if (index === 0)
        {
            add(value);
            return true;
        }
        // if index > 0 in an empty list, add item at the head
        else if ((index !== 0) && (this.isEmpty()))
        {
            add(value);
            return true;;
        }
        
        let iter = this._head.next;
        let i = 0;
        
        while (iter !== this._head)
        {
            if (i === index)
            {
                let parent = iter.previous;
                
                let node = new Node(value);
                
                node.next = iter;
                node.previous = parent;
                iter.previous = node;
                parent.next = node;
                
                this._length += 1;
                
                return true;
            }
            
            i += 1;
            iter = iter.next;
        }
        
        return false;
    }
    
    remove(value)
    {
        let iter = this._find(value);
        
        // item found
        if (iter !== null)
        {
            let child = iter.next;
            let parent = iter.previous;
            
            parent.next = child;
            child.previous = parent;
            iter = null;
            
            this._length -= 1;
            
            return true;
        }
        
        return false;
    }
    
    add(value)
    {
        let node = new Node(value);
            
        if (this.isEmpty())
        {
            this._head.next = node;
            this._head.previous = node;
            node.previous = this._head;
            node.next = this._head;
        }
        else
        {
            let head = this._head.next;
            
            this._head.next = node;
            node.next = head;
            head.previous = node;
            node.previous = this._head;
        }
        
        this._length += 1;
    }
    
    addRange(collection)
    {
        collection.forEach( item => {
            this.add(item);
        });
    }
    
    
    print()
    {
        // empty list
        if (this.isEmpty())
        {
            console.log("[ ]");
            return;
        }
        
        let iter = this._head.next;
        let stringBuilder = "[";
        
        while (iter !== this._head) 
        {
            stringBuilder += " ";
            stringBuilder += iter.data;
            stringBuilder += ",";
            iter = iter.next;
        }
        
        let regex = /,([^,]*)$/;
        stringBuilder = stringBuilder.replace(regex, " ]");
        
        
        console.log(stringBuilder);
    }
};

File: Algo Java/datastructures/CircularLinkedList.java
/**
 *
 * @author Gaurab R. Gautam
 */
package datastructures;

/**
 *
 * @author Anonymous
 * @param <T>
 */
public class CircularLinkedList<T> 
{
    private Node<T> head = null;
    private int size = 0;
    
    public CircularLinkedList()
    {
        this.head = new Node();
        this.head.value = null;
    }
    
    public CircularLinkedList(int iniSize)
    {
        this.head = new Node();
        this.head.value = null;
        T val = null;
        
        for (int i = 0; i < iniSize; i++)
        {
            Node<T> node = new Node(null);
        
            if (size == 0)
            {
                this.head.next = node;
                this.head.previous = node;
                node.previous = head;
                node.next = head;
            }
            else
            {
                Node<T> first = this.head.next;
                this.head.next = node;
                node.previous = this.head;
                first.previous = node;
                node.next = first;
            }
            
            this.size += 1;
        }
        //System.out.println("size = " + size + " : " + this);
    }
    
    public int size()
    {
        return this.size;
    }
    
    public void clear()
    {
        for (int i = 0; i < this.size; i++)
        {
            delete(get(i));
        }
        
        this.head = new Node();
        this.size = 0;
    }
    
    public T get(int index)
    {
        int i = 0;
        Node iter = head.next;
        
        while (i != index && i < this.size)
        {
            iter = iter.next;
            i += 1;
        }
        
        return (T)iter.value;
    }
    
    public void set(int index, T val)
    {
        int i = 0;
        Node iter = head.next;
        
        while (i != index && i < this.size)
        {
            iter = iter.next;
            i += 1;
        }
        
        iter.value = val;
    }
    
    
    public boolean isEmpty()
    {
        return (this.size == 0);
    }
    
    public void insert(T val)
    {
        Node<T> node = new Node(val);
        
        if (this.isEmpty())
        {
            this.head.next = node;
            this.head.previous = node;
            node.previous = head;
            node.next = head;
        }
        else
        {
            Node<T> first = this.head.next;
            this.head.next = node;
            node.previous = this.head;
            first.previous = node;
            node.next = first;
        }
            
        this.size += 1;
    }
    
    public void insert(T val, int index)
    {
        int i = 0;
        Node iter = head.next;
        
        while (i != index && i < this.size)
        {
            iter = iter.next;
            i += 1;
        }
        
        iter.value = val;
    }
    
    public boolean search (T val)
    {
        return (find(val) != null);
    }
    
    private Node<T> find (T val)
    {
        Node<T> iter = this.head.next;
        
        while (iter != this.head && !iter.value.equals(val))
        {
            iter = iter.next;
        }
        
        if (iter == this.head)
        {
            //System.out.println(val + " not found!");
            iter = null;
        }
        
        return iter;
    }
    
    public boolean delete(T val)
    {
        Node<T> node = find(val);
        
        if (node != null)
        {
            Node<T> parent = node.previous;
            Node<T> child = node.next;
            
            parent.next = node.next;
            child.previous = node.previous;
            
            node = null;
            this.size -= 1;
        }
        
        return false;
    }
    
    public void print()
    {
        System.out.println(toString());
    }
    
    @Override
    public String toString()
    {
       StringBuilder str = new StringBuilder();
       
        Node<T> iter = this.head.next;
        str.append("[");
        
        while (iter != this.head)
        {
            str.append(iter.value).append(iter.next != head ? ", " : "");
            iter =  iter.next;
        }
        
        str.append("]");
        
        return str.toString();
    }
}
